source/common/ssl/context_config_impl.cc

```diff
@@ -1,17 +1,43 @@
 #include "common/ssl/context_config_impl.h"
 
+#include <memory>
 #include <string>
 
 #include "common/common/assert.h"
+#include "common/common/empty_string.h"
 #include "common/config/datasource.h"
 #include "common/config/tls_context_json.h"
 #include "common/protobuf/utility.h"
+#include "common/ssl/tls_certificate_config_impl.h"
 
 #include "openssl/ssl.h"
 
 namespace Envoy {
 namespace Ssl {
 
+namespace {
+
+std::string readConfig(
+    const envoy::api::v2::auth::CommonTlsContext& config, Secret::SecretManager& secret_manager,
+    const std::function<std::string(const envoy::api::v2::auth::TlsCertificate& tls_certificate)>&
+        read_inline_config,
+    const std::function<std::string(const Ssl::TlsCertificateConfigImpl& secret)>& read_secret) {
+  if (!config.tls_certificates().empty()) {
+    return read_inline_config(config.tls_certificates()[0]);
+  } else if (!config.tls_certificate_sds_secret_configs().empty()) {
+    auto name = config.tls_certificate_sds_secret_configs()[0].name();
+    auto secret = secret_manager.findSecret(Secret::Secret::SecretType::TLS_CERTIFICATE, name);
+    if (!secret) {
+      throw EnvoyException(fmt::format("Static secret is not defined: {}", name));
+    }
+    return read_secret(*(std::dynamic_pointer_cast<Ssl::TlsCertificateConfigImpl>(secret)).get());
```

Comment:
nit: `dynamic_cast<const Ssl::TlsCertificateConfigImpl&>(*secret)` should work.

Reply:
I would rather just get rid of the dynamic casts. I don't think it's necessary. Since @PiotrSikora hasn't responded back, can we please rework this so that the secret manager supports different types of secrets with a normal class structure? I think that will be the cleanest moving forward. Thank you!

Reply:
I will update. Thanks.

Reply:
Removed dynamic casts